src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -0,0 +1,185 @@
+/// config templates.
+#[cfg(target_arch = "x86_64")]
+pub mod x86_64 {
+    use std::str::FromStr;
+
+    use serde::de::Error as SerdeError;
+    use serde::{Deserialize, Deserializer};
+
+    use crate::guest_config::templates::deserialize_u64_from_str;
+
+    /// CPUID register enumeration
+    #[allow(unused)]
+    #[allow(missing_docs)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub enum CpuidRegister {
+        Eax,
+        Ebx,
+        Ecx,
+        Edx,
+    }
+
+    /// Composite type that holistically provides
+    /// the location of a specific register being used
+    /// in the context of a CPUID tree.
+    #[allow(unused)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct CpuidRegisterId {
+        leaf: u32,
+        subleaf: u32,
+        flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        pointer: CpuidRegister,
+    }
+
+    /// Wrapper type to containing x86_64 CPU config modifiers.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct X86_64CpuTemplate {
+        /// Modifiers for CPUID configuration.
+        // pub cpuid_modifiers: Vec<RegisterBitMask>,
+        /// Modifiers for model specific registers.
+        pub msr_modifiers: Vec<RegisterModifier>,
+    }
+
+    /// Bit-mask based wrapper to apply
+    /// changes to a given register's value
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct RegisterModifier {
+        /// Pointer of the location to be bit masked.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        addr: u32,
+        /// Mask to be applied to register's value at the address
+        /// provided.
+        mask: BitMask,
+    }
+
+    fn deserialize_u32_from_str<'de, D>(deserializer: D) -> Result<u32, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let number_str = String::deserialize(deserializer)?;
+        let deserialized_number: u32 = if number_str.len() > 2 {
+            match &number_str[0..2] {
+                "0b" => u32::from_str_radix(&number_str[2..], 2),
+                "0x" => u32::from_str_radix(&number_str[2..], 16),
+                _ => u32::from_str(&number_str),
+            }
+            .map_err(|err| {
+                SerdeError::custom(format!(
+                    "Failed to parse string [{}] as a number for CPU template - {:?}",
+                    number_str, err
+                ))
+            })?
+        } else {
+            u32::from_str(&number_str).map_err(|err| {
+                SerdeError::custom(format!(
+                    "Failed to parse string [{}] as a decimal number for CPU template - {:?}",
+                    number_str, err
+                ))
+            })?
+        };
+
+        Ok(deserialized_number)
+    }
+
+    /// Bit mask to adjust targeted bits to the value,
+    /// applied by the filter only.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct BitMask {
```

Comment:
I think the `RegisterModifier` already describes what you propose akin to the operation? It contains a target address and a mask to be applied. Value seems appropriate to me because it is the value that we want the result to be, and the filter decides which bits from the value are relevant.

If you are not comfortable with the name `value` in this context we can change that?

Reply:
No, I'm good with the `value`. What I don't like is the `value` is a part of the `BitMask`. A modifier contains both a `value` and a `mask`. A `mask` should not contain a `value`, they are complementary to each other.